crypto/fipsmodule/rsa/rsa_impl.c

```diff
@@ -1327,37 +1327,46 @@ static int RSA_generate_key_ex_maybe_fips(RSA *rsa, int bits,
   RSA *tmp = NULL;
   uint32_t err;
   int ret = 0;
+  int failures;
+  int num_attempts = 0;
 
+  do {
   // |rsa_generate_key_impl|'s 2^-20 failure probability is too high at scale,
-  // so we run the FIPS algorithm |MAX_RSA_KEYGEN_ATTEMPTS| times,
-  // bringing it down to 2^-80 when it's equal to 4. We
+  // so we run the FIPS algorithm four times, bringing it down to 2^-80. We
   // should just adjust the retry limit, but FIPS 186-4 prescribes that value
   // and thus results in unnecessary complexity.
-  int failures = 0;
-  do {
-    ERR_clear_error();
-    // Generate into scratch space, to avoid leaving partial work on failure.
-    tmp = RSA_new();
-    if (tmp == NULL) {
-      goto out;
-    }
+  failures = 0;
+    do {
+      ERR_clear_error();
+      // Generate into scratch space, to avoid leaving partial work on failure.
+      tmp = RSA_new();
+      if (tmp == NULL) {
+        goto out;
+      }
 
-    if (rsa_generate_key_impl(tmp, bits, e_value, cb)) {
-      if (check_fips && !RSA_check_fips(tmp)) {
-        failures++;
-      } else {
+      if (rsa_generate_key_impl(tmp, bits, e_value, cb)) {
         break;
       }
-    }
 
-    err = ERR_peek_error();
-    RSA_free(tmp);
-    tmp = NULL;
+      err = ERR_peek_error();
+      RSA_free(tmp);
+      tmp = NULL;
+      failures++;
+
+      // Only retry on |RSA_R_TOO_MANY_ITERATIONS|. This is so a caller-induced
+      // failure in |BN_GENCB_call| is still fatal.
+    } while (failures < 4 && ERR_GET_LIB(err) == ERR_LIB_RSA &&
```

Comment:
any particular reason to replace MAX_RSA_KEYGEN_ATTEMPTS with 4?

Reply:
4 was originally there, and it makes the probability of failure of `rsa_generate_key_impl` down from 2^{-20} to 2^{-80} as per the original comment. The inner loop is exactly as it was before the PCT iterate-and-abort changes and I think of it as one step of RSA key generation.